x-pack/plugins/apm/server/routes/backends/get_top_backend_spans.ts

```diff
@@ -0,0 +1,139 @@
+  SERVICE_ENVIRONMENT,
+  SERVICE_NAME,
+  SPAN_DESTINATION_SERVICE_RESOURCE,
+  SPAN_DURATION,
+  SPAN_NAME,
+  TRACE_ID,
+  TRANSACTION_ID,
+  TRANSACTION_NAME,
+  TRANSACTION_TYPE,
+} from '../../../common/elasticsearch_fieldnames';
+import { Environment } from '../../../common/environment_rt';
+import { ProcessorEvent } from '../../../common/processor_event';
+import { environmentQuery } from '../../../common/utils/environment_query';
+import { AgentName } from '../../../typings/es_schemas/ui/fields/agent';
+import { Setup } from '../../lib/helpers/setup_request';
+
+const MAX_NUM_SPANS = 1000;
+
+export interface BackendSpan {
+  '@timestamp': number;
+  spanName: string;
+  serviceName: string;
+  agentName: AgentName;
+  traceId: string;
+  transactionId?: string;
+  transactionType?: string;
+  transactionName?: string;
+  duration: number;
+}
+
+export async function getTopBackendSpans({
+  setup,
+  backendName,
+  spanName,
+  start,
+  end,
+  environment,
+  kuery,
+}: {
+  setup: Setup;
+  backendName: string;
+  spanName: string;
+  start: number;
+  end: number;
+  environment: Environment;
+  kuery: string;
+}): Promise<BackendSpan[]> {
+  const { apmEventClient } = setup;
+
+  const spans = (
+    await apmEventClient.search('get_top_backend_spans', {
+      apm: {
+        events: [ProcessorEvent.span],
+      },
+      body: {
+        size: MAX_NUM_SPANS,
+        query: {
+          bool: {
+            filter: [
+              ...rangeQuery(start, end),
+              ...environmentQuery(environment),
+              ...kqlQuery(kuery),
+              ...termQuery(SPAN_DESTINATION_SERVICE_RESOURCE, backendName),
+              ...termQuery(SPAN_NAME, spanName),
+            ],
+          },
+        },
+        _source: [
+          TRACE_ID,
+          TRANSACTION_ID,
+          SPAN_NAME,
+          SERVICE_NAME,
+          SERVICE_ENVIRONMENT,
+          AGENT_NAME,
+          SPAN_DURATION,
+          '@timestamp',
+        ],
+      },
+    })
+  ).hits.hits.map((hit) => hit._source);
+
+  const transactionIds = compact(spans.map((span) => span.transaction?.id));
+
+  const transactions = (
+    await apmEventClient.search('get_transactions_for_backend_spans', {
+      apm: {
+        events: [ProcessorEvent.transaction],
+      },
+      body: {
+        size: transactionIds.length,
+        query: {
+          bool: {
+            filter: [...termsQuery(TRANSACTION_ID, ...transactionIds)],
+          },
+        },
+        _source: [TRANSACTION_ID, TRANSACTION_TYPE, TRANSACTION_NAME],
+      },
+    })
+  ).hits.hits.map((hit) => hit._source);
+
+  const transactionsById = keyBy(
+    transactions,
+    (transaction) => transaction.transaction.id
+  );
+
+  return spans.map((span): BackendSpan => {
+    const transaction = span.transaction
+      ? transactionsById[span.transaction.id]
+      : undefined;
+
+    return {
+      '@timestamp': new Date(span['@timestamp']).getTime(),
+      spanName: span.span.name,
```

Comment:
Is `spanName` needed?

Reply:
It's not needed, as it's presumed by the query parameter. However, I ran into a bug when writing API tests, because I didn't add a query for `span.name`. Adding `span.name` makes it easier to write tests. It was also in the original design. I'm fine with removing it, but leaning slightly towards keeping it because of the API tests.

Reply:
It's an extra property to be sent in the response but it's not a big deal for me either